### ISSUE TITLE
map with different length iterators

### ISSUE THREAD

**Issue by boathit** [JuliaLang/julia#13361]
Is it reasonable that `map` can deal with infinite length iterators well but fails to work with two different finite length iterators?

```
map(+, [1,2,3], repeated(5))

3-element Array{Any,1}:
 6
 7
 8


map(+, [1,2,3], repeated(5, 10))

3-element Array{Any,1}:
 6
 7
 8


map(+, 1:3, repeated(5, 10))

3-element Array{Any,1}:
 6
 7
 8


```

works well

```
 map(+, [1,2,3], 1:10)

ERROR: DimensionMismatch("dimensions must match")
 in map at abstractarray.jl:1310

map(+, 1:3, 1:10)

ERROR: DimensionMismatch("dimensions must match")
 in map at abstractarray.jl:1310


```

error occurs

**Comment by ihnorton** [comment]
Suggested alternative:
```
map(+, [1,2,3], 1:10...)
```

I'm unsure if it would be a good idea to auto-expand the range `1:10` object (a.k.a. `colon(1,10)`) in cases like this.

(there are a some similar cases where such auto-expansion might be convenient -- but possibly undesirable globally -- such as indexing a KeyIterator: `keys(Dict(1=>"a"))[1]` -> `MethodError`. instead requires:  `[keys(Dict(1=>"a"))...][1]`)

**Comment by boathit** [comment]
`map(+, [1,2,3], 1:10...)` will be expanded into `map(+, [1,2,3], 1,2,3,4,5,6,7,8,9,10)` which departs from the original intention that expects to perform `map(+, [1,2,3], [1,2,3,4,5,6,7,8,9,10])`

**Comment by JeffBezanson** [comment]
Maybe this should be allowed for 1-d arrays of different lengths?

**Comment by raghav9-97** [comment]
@JeffBezanson Is it logically correct if map(f, x, y) work with different lengths of x and y and if length(x) < length(y) the resulting array size should be length(x)?

**Comment by raghav9-97** [comment]
I also think that issue #29523 is related to this problem.

**Comment by timholy** [comment]
An orthogonal solution is to define new iterators like `repeatlast(x)` which turns a container `x` into an infinite-length iterator, "padding" with the final element. I think this would be much more explicit about intent, and also more flexible.

**Comment by raghav9-97** [comment]
The ```promote_shape(::Indices, ::Indices)``` function works just opposite to what has been suggested for Iterators.truncate(y, x).So why we can't just change the promote_shape function to match a longer iterator to shorter one?

**Comment by raghav9-97** [comment]
I think the map function should call truncate(y, x) instead of promote_shape directly for truncating longer iterator length.What do you say?

**Comment by raghav9-97** [comment]
The map function makes a call to promote_shape which functions differently while we need to compare the two array size and return one with smaller size.

**Comment by raghav9-97** [comment]
I think truncate function can be applied as truncate(y::Indices, x::Indices) then compare both the indices as the case with promote_shape and return shorter iterator and replacing promote_shape call with truncate in map.

**Comment by jonas-schulze** [comment]
IIUC this should be fixed by #29927:

```
map(f, iters...) = collect(Generator(f, iters...))
Generator(f, I1, I2, Is...) = Generator(a->f(a...), zip(I1, I2, Is...))
```

**Comment by mschauer** [comment]
Yes, the reported issue was fixed by that:
```
julia> map(+, [1,2,3], 1:10)
3-element Array{Int64,1}:
 2
 4
 6
```
